# Importamos las librerías

In [110]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import RidgeClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# Cargamos el dataset

In [111]:
data = pd.read_csv("C:/Users/Daniel/Downloads/cancer_classification.csv")
data

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  \
0                  0.2654          0.4601                  0.11890   
1                  0.1860          0.2750                  0.08902   
2                  0.2430          0.3613                  0.08758   
3                  0.2575          0.6638                  0.17300   
4                  0.1625          0.2364                  0.07678   
..                    ...    

# Identificamos tipos de datos de dataset

In [112]:
data.dtypes

mean radius                float64
mean texture               float64
mean perimeter             float64
mean area                  float64
mean smoothness            float64
mean compactness           float64
mean concavity             float64
mean concave points        float64
mean symmetry              float64
mean fractal dimension     float64
radius error               float64
texture error              float64
perimeter error            float64
area error                 float64
smoothness error           float64
compactness error          float64
concavity error            float64
concave points error       float64
symmetry error             float64
fractal dimension error    float64
worst radius               float64
worst texture              float64
worst perimeter            float64
worst area                 float64
worst smoothness           float64
worst compactness          float64
worst concavity            float64
worst concave points       float64
worst symmetry      

# Realizamos conteo de valores nulos

In [113]:
data.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
benign_0__mal_1            0
dtype: int64

In [114]:
#Conteo de la columna -benign_0__mal_1- 0 = Benigno 1 = Maligno

In [115]:
data["benign_0__mal_1"].value_counts()

1    357
0    212
Name: benign_0__mal_1, dtype: int64

## Seleccionamos datos para el eje X y Y

In [116]:
x = data.drop("benign_0__mal_1", axis=1)
y = data["benign_0__mal_1"]

## Normalización de funciones

In [117]:

scaler = MinMaxScaler()
# transform data
scaled = scaler.fit_transform(x)

In [118]:

x_train, x_test, y_train, y_test = train_test_split(scaled, y, test_size=0.2, random_state=0)

## Modelos utilizados

### Modelo lineal

In [119]:

lin_class = RidgeClassifier()
lin_class.fit(x_train, y_train)

RidgeClassifier()

In [120]:
y_preds = lin_class.predict(x_test)

In [121]:


print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        47
           1       0.93      1.00      0.96        67

    accuracy                           0.96       114
   macro avg       0.97      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



In [122]:
confusion_matrix(y_test, y_preds)

array([[42,  5],
       [ 0, 67]], dtype=int64)

#### En base a la correlación de la muestra de los datos de prueba y las predicciones aplicadas aplicando el modelo lineal tenemos 42 y 67 lo cual es bastante bueno

In [123]:
lin_class.score(x_test, y_test)

0.956140350877193

#### El modelo lineal tiene una efectividad del 95% lo cual aparentemente sería el mejor modelo a ser utilizado

### Árbol de decisión

In [124]:

dtc = DecisionTreeClassifier()
dtc.fit(x_train, y_train)

DecisionTreeClassifier()

In [125]:
y_preds_dtc = dtc.predict(x_test)

In [126]:
print(classification_report(y_test,y_preds_dtc))

              precision    recall  f1-score   support

           0       0.85      0.96      0.90        47
           1       0.97      0.88      0.92        67

    accuracy                           0.91       114
   macro avg       0.91      0.92      0.91       114
weighted avg       0.92      0.91      0.91       114



In [127]:
confusion_matrix(y_test, y_preds_dtc)

array([[45,  2],
       [ 8, 59]], dtype=int64)

#### En base a la correlación de la muestra de los datos de prueba y las predicciones aplicadas aplicando el modelo de árboles de decisión tenemos 44 y 59 lo cual es un rango aceptable

In [128]:
dtc.score(x_test,y_test)

0.9122807017543859

#### El modelo de Árboles de decisión tienen una efectividad del 91% lo cual nos hace presumir que debemos utilizar otro modelo para mejorar este valor

### Random Forest 

In [129]:

rfc = RandomForestClassifier()
rfc.fit(x_train, y_train)

RandomForestClassifier()

In [130]:
y_preds_rfc = rfc.predict(x_test)

In [131]:
print(classification_report(y_test, y_preds_rfc))

              precision    recall  f1-score   support

           0       0.96      0.96      0.96        47
           1       0.97      0.97      0.97        67

    accuracy                           0.96       114
   macro avg       0.96      0.96      0.96       114
weighted avg       0.96      0.96      0.96       114



In [132]:
confusion_matrix(y_test, y_preds_rfc)

array([[45,  2],
       [ 2, 65]], dtype=int64)

#### En base a la correlación de la muestra de los datos de prueba y las predicciones aplicadas aplicando el modelo random forest tenemos 46 y 66 lo cual es un valor bastante alto pero la efefctividad del modelo tendrá la última palabra

In [135]:
rfc.score(x_test, y_test)

0.9649122807017544

#### El modelo Random Forest tiene una efectividad del 96% lo cual nos indica un sobreajuste 